presto-rcfile/src/main/java/com/facebook/presto/rcfile/RcFileDecoderUtils.java

```diff
@@ -102,6 +112,39 @@ private static long readVIntInternal(Slice slice, int start, int length)
         return isNegativeVInt(slice.getByte(start)) ? ~value : value;
     }
 
+    public static long findFirstSyncPosition(RcFileDataSource dataSource, long offset, long length, long syncFirst, long syncSecond)
+            throws IOException
+    {
+        requireNonNull(dataSource, "dataSource is null");
+        checkArgument(offset >= 0, "offset is negative");
+        checkArgument(length >= 1, "length must be at least 1");
+
+        Slice sync = Slices.allocate(SIZE_OF_INT + SIZE_OF_LONG + SIZE_OF_LONG);
+        sync.setInt(0, 0xFFFF_FFFF);
+        sync.setLong(SIZE_OF_INT, syncFirst);
+        sync.setLong(SIZE_OF_INT + SIZE_OF_LONG, syncSecond);
+
+        // read 1 MB chunks at a time, but only skip ahead 1 MB - SYNC_SEQUENCE_LENGTH bytes
+        // this causes a re-read of SYNC_SEQUENCE_LENGTH bytes each time, but is much simpler code
+        byte[] buffer = new byte[Ints.checkedCast(min(1 << 20, length + SYNC_SEQUENCE_LENGTH))];
+        Slice bufferSlice = Slices.wrappedBuffer(buffer);
+        for (long position = 0; position < length; position += bufferSlice.length() - SYNC_SEQUENCE_LENGTH) {
+            // either fill the buffer entirely, or read enough to allow all bytes in offset + length to be a start sequence
+            int bufferSize = Ints.checkedCast(min(buffer.length, length + SYNC_SEQUENCE_LENGTH - position));
+            // don't read off the end of the file
+            bufferSize = Ints.checkedCast(min(bufferSize, dataSource.getSize() - offset - position));
+
+            dataSource.readFully(offset + position, buffer, 0, bufferSize);
+
+            int index = bufferSlice.indexOf(sync);
+            if (index >= 0 && index + position < length) {
```

Comment:
it is possible (but unlikely) to get a sync point off the end of the "length-region", because we read chunks and the last chunk may not be aligned to the end of the "length-region".  This means that the new read, plus dirty data from the previous read could look like a sync point (although really unlikely).